lib/src/gzip_encoding.dart

```diff
@@ -0,0 +1,230 @@
+    bool addCompressionRatioHeader = true}) {
+  return (Handler innerHandler) {
+    return (request) {
+      if (!acceptsGzipEncoding(request)) {
+        return innerHandler(request);
+      }
+      return Future.sync(() => innerHandler(request)).then((response) =>
+          gzipEncodeResponse(response,
+              compressionLevel: compressionLevel,
+              addCompressionRatioHeader: addCompressionRatioHeader));
+    };
+  };
+}
+
+/// Returns `true` if the request accepts `gzip` encoding.
+bool acceptsGzipEncoding(Request request) {
+  var acceptEncoding = request.headers[HttpHeaders.acceptEncodingHeader];
+  return acceptEncoding?.contains('gzip') ?? false;
+}
+
+/// Converts [response] to a `gzip` encoding response.
+/// Checks [canGzipEncodeResponse].
+///
+/// See [createGzipMiddleware].
+FutureOr<Response> gzipEncodeResponse(Response response,
+    {int compressionLevel = defaultGzipEncodingCompressionLevel,
+    bool addCompressionRatioHeader = true}) async {
+  if (!canGzipEncodeResponse(response)) {
+    return response;
+  }
+
+  var bufferInitialCapacity = response.contentLength ?? 1024 * 4;
+
+  // Read the body bytes from the response:
+  var bytesBuffer = await response.read().fold<_BytesBuffer>(
+      _BytesBuffer(bufferInitialCapacity),
+      (result, bytes) => result..addAll(bytes));
+
+  var gzipEncoder = compressionLevel == defaultGzipEncodingCompressionLevel
+      ? _defaultGzipEncoder
+      : ZLibEncoder(gzip: true, level: compressionLevel);
+
+  // Compressed body:
+  var compressedBody = gzipEncoder.convert(bytesBuffer.toUint8List());
+
+  var bodyLength = bytesBuffer.length;
+  var compressedBodyLength = compressedBody.length;
+  var compressionRatio = compressedBodyLength / bodyLength;
+
+  var headers = Map<String, String>.from(response.headers);
+
+  headers[HttpHeaders.contentEncodingHeader] = 'gzip';
+  headers[HttpHeaders.contentLengthHeader] = compressedBodyLength.toString();
+  headers['X-Compression-Ratio'] =
+      '$compressionRatio ($compressedBodyLength/$bodyLength)';
+
+  return response.change(headers: headers, body: compressedBody);
+}
+
+/// Returns `true` if [response] can be compressed.
+///
+/// Checks:
+/// - `Content-Encoding`: if already present, can't change it to `gzip`.
+/// - `Content-Type`: checks if [isAlreadyCompressedContentType].
+/// - `Content-Length`: checks if too small for compression (< 512).
+bool canGzipEncodeResponse(Response response) {
+  var headerContentEncoding =
+      response.headers[HttpHeaders.contentEncodingHeader];
+
+  // If the response already defines a `Content-Encoding` header it
+  // won't apply the the `gzip` encoding to preserve the response behavior.
+  if (headerContentEncoding != null && headerContentEncoding.isNotEmpty) {
+    return false;
+  }
+
+  var responseContentLength = response.contentLength;
+
+  // A small body will not benefit from being compressed:
+  if (responseContentLength != null && responseContentLength < 512) {
```

Comment:
Allow this constant to be overridden?

Reply:
Yes, but users might want to make this larger.

Reply:
I have added the parameter `minimalGzipContentLength` to allow that.